fix(rich-text-editor): NO-JIRA multiple improvements

braddialpad opened the pull request:
# fix(rich-text-editor): multiple improvements
<!--- Feel free to remove any unused sections -->

## Obligatory GIF (super important!)

<!--- go to giphy.com, pick a gif, click share, copy gif link, paste within the () brackets below. -->

![Obligatory GIF](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExeTNuZmQ5MDFiNmlzcm9wcm0xZzYxOHFpOXczcnRldm1ycXY5ejA5ZSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/FY8c5SKwiNf1EtZKGs/giphy.gif)

## :hammer_and_wrench: Type Of Change

<!--- Tick or place an `x` for the type of change. Should match the type in the commit message / PR title
in https://github.com/dialpad/dialtone/blob/staging/.github/COMMIT_CONVENTION.md -->

These types will increment the version number on release:

- [x] Fix

## :book: Jira Ticket

https://dialpad.atlassian.net/browse/DLT-1825
https://dialpad.atlassian.net/browse/DLT-1858
https://dialpad.atlassian.net/browse/DLT-1856
https://dialpad.atlassian.net/browse/DLT-1674

## :book: Description

Fixes for multiple reported issues with rich-text-editor / message-input

- DLT-1825: prevent the triggering of keydown events in suggestion popover when the item count is 0, this was causing multiple errors
- DLT-1858, DLT-1856: trigger enter event from the input field directly rather than the input wrapper.
- DLT-1674: Designer reported UI improvements. Lots of these will be fixed when nina's scrollbar PR gets merged. I just changed the button colors as requested.

## :bulb: Context

Reported issues from the implementation of message input into dialpad.

## :pencil: Checklist

<!--- Tick or place an `x` in all of the checkboxes that apply -->
<!--- Remove checkboxes that do not apply -->

For all PRs:

- [x] I have ensured no private Dialpad links or info are in the code or pull request description (Dialtone is a public repo!).
- [x] I have reviewed my changes.
- [x] I have added all relevant documentation.
- [x] I have considered the performance impact of my change.

For all Vue changes:

- [x] I have made my changes in Vue 2 and Vue 3. Note: you may sync your changes from Vue 2 to Vue 3 (or vice versa) using the `./scripts/dialtone-vue-sync.sh` script. Read docs here: [Dialtone Vue Sync Script](../packages/dialtone-vue3/.github/CONTRIBUTING.md#dialtone-vue-sync-script)

For all CSS changes:

- [x] I have used design tokens whenever possible.

## :crystal_ball: Next Steps

Notify the messaging team that these have been fixed.

braddialpad committed:
fix(rich-text-editor): DLT-1856 emit enter event from rich-text

braddialpad committed:
vue 3

braddialpad committed:
don't trigger keydown if length is 0

braddialpad committed:
vue 3

braddialpad committed:
buttons to foreground muted

braddialpad committed:
vue3

iropolo commented:
On Conversation View -> Message input -> Default vue2 if you press enter, it doesn't add a jump line, it does it in vue3

https://github.com/dialpad/dialtone/assets/89984179/21181c45-cefe-42c2-8640-08efc07bf5ad


Non on vue3 if you press arrow down key it trigger console error.

https://github.com/dialpad/dialtone/assets/89984179/a36a8204-fea0-41bf-9d92-f06954ee0f47

iropolo commented on packages/dialtone-vue2/components/rich_text_editor/extensions/suggestion/SuggestionList.vue:
```diff
@@ -81,9 +82,7 @@ export default {
       }
 
       if (event.key === 'Enter' || event.key === 'Tab') {
-        event.preventDefault(); // prevent moving the focus for the tab key
         this.selectHandler();
-        event.stopPropagation();
```
not sure why these were useful

braddialpad replied:
yeah the event in the template already had `prevent.stop` on it, haha

iropolo reviewed (approved):
Apart from my comments, fixs looks good

braddialpad commented:
> On Conversation View -> Message input -> Default vue2 if you press enter, it doesn't add a jump line, it does it in vue3

It's actually only supposed to add a line if you use "shift + enter" pressing enter normally would send the message. I will look into this.

braddialpad commented:
Couple more small requested fixes that came in from design today that I will add and document.

braddialpad committed:
add tertiary subtitle

braddialpad committed:
vue 3

braddialpad committed:
set max height on dialog

braddialpad committed:
vue 3

braddialpad committed:
add top slot

braddialpad committed:
vue 3

braddialpad commented:
Added: 
- tertiary foreground color to slash command subtitle
- moved max-height to the dialog rather than the child container, so it is retained when scrolling to bottom.
- slot for "top" so they can add custom messages above the input field, but still contained within the borders.

iropolo commented:
> > On Conversation View -> Message input -> Default vue2 if you press enter, it doesn't add a jump line, it does it in vue3
> 
> It's actually only supposed to add a line if you use "shift + enter" pressing enter normally would send the message. I will look into this.

K, seems like in vue3 it sends the actions and also adds a jump line.

iropolo commented on packages/dialtone-vue2/components/rich_text_editor/extensions/suggestion/SuggestionList.vue:
```diff
@@ -70,6 +70,7 @@ export default {
 
   methods: {
     onKeyDown ({ event }) {
+      if (this.items.length === 0) return;
```
This seems like not working, I can still trigger errors in console pressing arrow keys

braddialpad replied:
> This seems like not working, I can still trigger errors in console pressing arrow keys

hmm can you tell me the steps you take to trigger this?

iropolo replied:
ahh, seems like `this.items` is not empty on a second search so it always pass this validation.

Check the video, after I type `@` I start to press the down arrow key.
Then I delete it and type again `@` and start to press down arrow key, console error appears.

https://github.com/dialpad/dialtone/assets/89984179/caf21634-7675-42e2-b51f-08dfce178343

iropolo replied:
There is another issue but it was there before.
If you press up arrow key it triggers a console error that seems to be related to destroy the popup (probably because you are moving out the @)

To reproduce type `@` and quickly up arrow key.

https://github.com/dialpad/dialtone/assets/89984179/14bf055b-d0a1-4590-9da2-c87c5e2b93eb

braddialpad replied:
> Check the video, after I type `@` I start to press the down arrow key.
> Then I delete it and type again `@` and start to press down arrow key, console error appears.

Ah yeah, seems to happen if you press the arrow key before the dialog is loaded

braddialpad replied:
The variables for the component and the popup in suggestion.js were not getting set to null after being destroyed. They still had an object in them even though they were destroyed which kept many things "alive" when they shouldn't have been.

Should be fixed now.

braddialpad committed:
fix keypress bugs

braddialpad committed:
vue 3

bianca-artola-dialpad commented:
Sorry for disrupting here, but I noticed some issues with the input when adding `:`, `@` or `#` and then enter

https://github.com/dialpad/dialtone/assets/144126346/63a461ca-524f-4567-a6b1-131596a8a3a0

probably we need to execute submit action on those cases?

braddialpad committed:
remove color override

braddialpad committed:
vue3

braddialpad commented:
> Sorry for disrupting here, but I noticed some issues with the input when adding `:`, `@` or `#` and then enter

Ah yeah, that's definitely a problem, thanks.

braddialpad committed:
fix additional errors

braddialpad committed:
vue 3

braddialpad commented:
> Sorry for disrupting here, but I noticed some issues with the input when adding `:`, `@` or `#` and then enter
> 
> probably we need to execute submit action on those cases?

@bianca-artola-dialpad should be fixed now.

braddialpad committed:
Merge branch 'staging' into fix/message-input-various

braddialpad committed:
fix visual test flakiness on hovercard

braddialpad committed:
vue 3

francisrupert commented:
UPDATE: oops, I guess I meant to address this to @braddialpad (never mind @bianca-artola-dialpad)

New tiny (I promise) improvement while you're in here:

Change...

```css
.dt-message-input--focused {
  ...
}
```

...to...

```css
.dt-message-input:focus-within {
  ...
}
```

This will remove the awkward blinking observed as `dt-message-input--focused` is toggled (view the attached screencast). This tiny change will ensure any amount of keyboard or mouse actions with `dt-message-input` retains the style (shadow and bold border). 

Oh, and I assume remove associated tests and the current toggling of `dt-message-input--focused`, assuming it's not used for anything other than styling the container.

Thanks much! Let me know if you've any questions.

https://github.com/user-attachments/assets/5634e4a9-0a2e-41ee-b2a1-dd074a9f0843

bianca-artola-dialpad commented:
> > Sorry for disrupting here, but I noticed some issues with the input when adding `:`, `@` or `#` and then enter
> > probably we need to execute submit action on those cases?
> 
> @bianca-artola-dialpad should be fixed now.

yeah, I cannot reproduce it anymore, thank you!

braddialpad committed:
fix focus flickering

braddialpad committed:
vue 3

braddialpad committed:
remove tests that are no longer necessary

braddialpad committed:
vue 3

braddialpad committed:
Merge branch 'staging' into fix/message-input-various

braddialpad committed:
fix enter override vue3

braddialpad committed:
vue2

braddialpad commented:
> K, seems like in vue3 it sends the actions and also adds a jump line.

Fixed now.
